File: js/entities/Particle.js
// js/entities/Particle.js
import { state } from '../state.js';
import { Vector2 } from '../vector2.js';

export class Particle {
  constructor(pos, vel, life = 40) {
    this.pos = pos.clone();
    this.vel = vel;
    this.life = life;
    // Stored separately from the live-decrementing life above, so the
    // alpha fade in draw() scales correctly regardless of what life
    // value was passed in — previously hardcoded to /40, which was
    // wrong for any particle NOT created with the default life.
    this.maxLife = life;
    // Any valid CSS color string (hex, rgb, hsl, named) — alpha is
    // applied separately via globalAlpha in draw() rather than baked
    // into this string, so it works regardless of format. Defaults to
    // the original hardcoded yellow, so any existing caller that
    // doesn't set this explicitly (e.g. Player.js's teleport
    // particles) looks exactly the same as before.
    //
    // NOTE: draw() previously never actually read this property at
    // all — it was hardcoded to rgba(255,255,0,...) regardless of what
    // callers set here, silently no-op-ing utils.js's
    // createDeathParticles' own color override. Fixed below.
    this.color = 'rgb(255, 255, 0)';
    this.radius = 2; // matches the original hardcoded draw size
    // Both default to "no effect," so any existing caller that doesn't
    // set these explicitly (every current caller) behaves exactly as
    // before.
    this.drag = 1;     // velocity multiplier applied per frame — 1 = constant velocity forever, <1 = decelerates over time
    this.growRate = 0; // radius change per frame — 0 = constant size, positive = expands over its lifetime
  }
  update() {
    const timeScale = state.timeScale;
    this.pos.add(this.vel.clone().multiply(timeScale));
    // Math.pow for the drag decay — see Asteroid.js's identical
    // reasoning for why an exponential per-frame rate needs to be
    // exponentiated by timeScale, not just multiplied.
    this.vel = this.vel.multiply(Math.pow(this.drag, timeScale));
    this.radius += this.growRate * timeScale;
    this.life -= timeScale;
  }
  draw() {
    const ctx = state.ctx;
    if (this.life <= 0) return;
    ctx.globalAlpha = this.life / this.maxLife;
    ctx.beginPath();
    ctx.arc(this.pos.x, this.pos.y, this.radius, 0, Math.PI * 2);
    ctx.fillStyle = this.color;
    ctx.fill();
    ctx.globalAlpha = 1;
  }
}

File: lua/entities/Particle.lua
-- lua/entities/Particle.lua
--
-- Port of js/entities/Particle.js
-- A single simple physics particle (position + velocity + fading
-- life), used for debris/death/explosion bursts. js/utils.js's
-- createParticles/createDeathParticles helpers that build batches of
-- these are ported inline wherever they're needed (see
-- lua/entities/Explosion.lua) rather than as a shared module, since
-- this project doesn't have a shared lua/utils.lua yet.

local state = require("lua.state")

local Particle = {}
Particle.__index = Particle

function Particle.new(pos, vel, life)
  life = life or 40
  local self = setmetatable({}, Particle)
  self.pos = pos:clone()
  self.vel = vel
  self.life = life
  self.maxLife = life
  -- Plain {r, g, b} (0-1), not a CSS color string — LÖVE has no CSS
  -- color parsing, unlike the original's fillStyle string. Defaults to
  -- the original's hardcoded yellow.
  self.color = { 1, 1, 0 }
  self.radius = 2
  -- Both default to "no effect," matching the original: 1 = constant
  -- velocity forever, 0 = constant size.
  self.drag = 1
  self.growRate = 0
  return self
end

function Particle:update()
  local timeScale = state.timeScale or 1
  self.pos:addScaled(self.vel, timeScale)
  self.vel:scale(self.drag ^ timeScale)
  self.radius = self.radius + self.growRate * timeScale
  self.life = self.life - timeScale
end

function Particle:draw()
  if self.life <= 0 then return end
  love.graphics.setColor(self.color[1], self.color[2], self.color[3], self.life / self.maxLife)
  love.graphics.circle("fill", self.pos.x, self.pos.y, self.radius)
  love.graphics.setColor(1, 1, 1, 1)
end

return Particle
